WhateverGreen/kern_igfx_memory.cpp

```diff
@@ -0,0 +1,209 @@
+	requiresPatchingGraphics = false;
+	requiresPatchingFramebuffer = true;
+}
+
+void IGFX::DVMTCalcFix::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	// Guard: Disable the patch if it is not available on the current Intel platforms
+	if (!available) {
+		SYSLOG("igfx", "DVMT: This fix is not available on the current platform and has been disabled.");
+		return;
+	}
+		
+	// Enable the fix if designated boot argument or device property is found
+	enabled = checkKernelArgument("-igfxdvmt");
+	if (!enabled)
+		enabled = info->videoBuiltin->getProperty("enable-dvmt-calc-fix") != nullptr;
+	if (!enabled)
+		return;
+	
+	// Guard: Wait for the device to be published in the plane, otherwise `read` will crash on macOS Big Sur
+	if (!WIOKit::awaitPublishing(info->videoBuiltin)) {
+		SYSLOG("igfx", "DVMT: Failed to wait for the graphics device to be published. The fix has been disabled.");
+		enabled = false;
+		return;
+	}
+	
+	// Read the DVMT preallocated memory set in BIOS from the GMCH Graphics Control field at 0x50 (PCI0,2,0)
+	// TODO: Lilu needs to be updated to define the enumeration case `kIOPCIConfigGraphicsControl`
+	auto gms = WIOKit::readPCIConfigValue(info->videoBuiltin, /*WIOKit::kIOPCIConfigGraphicsControl*/ 0x50, 0, 16) >> 8;
+	
+	// Disable the fix if the GMS value can be calculated by Apple's formula correctly
+	// Reference: 10th Generation Intel Processor Families: Datasheet, Volume 2, Section 4.1.28
+	// https://www.intel.com/content/www/us/en/products/docs/processors/core/10th-gen-core-families-datasheet-vol-2.html
+	if (gms < 0x10) {
+		dvmt = gms * 32;
+		enabled = false;
+		DBGLOG("igfx", "DVMT: GMS value is supported by the driver. The fix has been disabled.");
+	} else if (gms == 0x20 || gms == 0x30 || gms == 0x40) {
+		dvmt = gms * 32;
+	} else if (gms >= 0xF0 && gms <= 0xFE) {
+		dvmt = ((gms & 0x0F) + 1) * 4;
+	} else {
+		SYSLOG("igfx", "DVMT: GMS value is reserved. Check your BIOS settings. DVMT will be set to 0.");
+		return;
+	}
+	
+	DBGLOG("igfx", "DVMT: GMS value is 0x%02x; DVMT pre-allocated memory is %d MB.", gms, dvmt);
+#ifdef DEBUG
+	info->videoBuiltin->setProperty("fw-dvmt-gms-field-value", gms, 8);
+	info->videoBuiltin->setProperty("fw-dvmt-preallocated-memory", dvmt, 32);
+#endif
+	dvmt *= (1024 * 1024); /* MB to Bytes */
+}
+
+void IGFX::DVMTCalcFix::processFramebufferKext(KernelPatcher &patcher, size_t index, mach_vm_address_t address, size_t size) {
+	// Find the address of `AppleIntelFramebufferController::FBMemMgr_Init()`
+	auto startAddress = patcher.solveSymbol(index, "__ZN31AppleIntelFramebufferController13FBMemMgr_InitEv", address, size);
+	if (!startAddress) {
+		SYSLOG("igfx", "DVMT: Failed to find FBMemMgr_Init().");
+		patcher.clearError();
+		return;
+	}
+	
+	//
+	// Locate Apple's formula: (GGCRegValue << 0x11) & 0xFE000000
+	//
+	// macOS Catalina 10.15.6 (19G2021)
+	// loc_3f61f: movl $0x50, %esi
+	// loc_3f624: call __ZN11IOPCIDevice20extendedConfigRead16Ey
```

Comment:
Can't we just use the hook of `extendedConfigRead` then, instead of patching the opcodes? WhateverGreen hooks  it anyway.

Reply:
No, we can't. The value read from PCI config space is valid but not compatible with Apple's formula. The value indicates the actual amount of DVMT memory preallocated by BIOS (e.g. 60MB), and we cannot fake it to pretend that we have allocated 64MB instead.

Reply:
What if you report a lower multiple of 32MB instead?

Reply:
Based on my experiment, 32MB is too small for some laptops, such as Surface Pro 7, with a high resolution display. There must be a reason why OEM chooses 60MB as a default value for DVMT preallocated memory on Ice Lake, so I think it's better to use their default value instead of trying to lower it down. (Well, if one could modify their BIOS settings via EFI shell or RU.EFI, they could always choose a higher value.)

Reply:
I see. Perhaps this could be implemented in OpenCore then? That would simplify the kernelspace part quite a bit...

Reply:
It cannot be.

Reply:
To document this a little, the problem is that DVMT config via the GMS register is impossible after the register gets locked for read-only access in PEI. Thus neither the bootloader nor the kernel can reconfigure it, and we must resort to teaching the driver to work with the given size. It may still be possible to provide a different size using OpRegion, but I wouldn't rely on that.

Reply:
@vit9696 Could this be implemented as a separate driver loaded from the "ConnectDrivers" section on OpenCore's config.plist?

Reply:
Instead of implementing it in OpenCore.efi main driver